Daylight: dropdown

GeorgePlukov opened the pull request.

GeorgePlukov commented on components/dropdown/README.md:
```diff
@@ -1,10 +1,12 @@
+A Dropdown is a button that opens a floating container to offer menu items or other content.
 
-### d2l-dropdown
+## Opener: Dropdown [d2l-dropdown]
```
Not sure how I feel about these titles, I thought some method of grouping openers/content would be ideal. Maybe these get split into a separate page?

margaree replied:
Yeah for sure - I like that you added something to distinguish the type. Can probably leave as is for now and see what Jeff thinks.

GeorgePlukov commented on components/dropdown/README.md:
```diff
@@ -1,10 +1,12 @@
 
 `d2l-dropdown` is a generic opener for dropdown content (`d2l-dropdown-content`, `d2l-dropdown-menu` or `d2l-dropdown-tabs`) enabling alternate opener implementation using existing elements/components. Provide and indicate your own opener element with the class attribute value `d2l-dropdown-opener`.  Wire-up is automatic.
 
+<!-- docs: demo live name:d2l-dropdown autoSize:false align:flex-start size:medium -->
```
The demos are currently static sized (`medium`) but the dropdown's start closed. @margaree To match what you did with date/time do you think that these should begin open (or maybe just the ones dealing with dropdown content)?

margaree replied:
Check with Jeff to be sure but I think we'd likely want to have them open as well and adopt a similar strategy as that page for how `opened` gets handled.

margaree commented on components/dropdown/README.md:
```diff
@@ -88,28 +85,33 @@ dropdown.addEventListener('click', function() {
+### Accessibility
 
-To make your `d2l-dropdown-button-subtle` accessible, use the following properties when applicable:
+To make your `d2l-dropdown-button-subtle` accessible, use the following properties when applicable
```
nit: probably want to keep the :

margaree commented on components/dropdown/README.md:
```diff
@@ -224,34 +231,35 @@ To make your usage of `d2l-dropdown-more` accessible, use the following property
-* `async resize()`: Call if the size of the content changes due to a change in
-a nested component. The nested component may choose to fire a custom event, which
-the component containing the `d2l-dropdown-content` can catch and call this method.
+### Methods
```
For a page I previously worked on Jeff had mentioned preferring leaving the methods as bullet points

GeorgePlukov replied:
Reverted

margaree commented on components/dropdown/README.md:
```diff
@@ -224,34 +231,35 @@ To make your usage of `d2l-dropdown-more` accessible, use the following property
+| `async resize()` | Call if the size of the content changes due to a change in a nested component. The nested component may choose to fire a custom event, which the component containing the `d2l-dropdown-content` can catch and call this method. |
 
-### d2l-dropdown-menu
+### Content: Menu [d2l-dropdown-menu]
```
I believe this should be an h2

margaree commented on components/dropdown/README.md:
````diff
@@ -284,11 +292,14 @@ the component containing the `d2l-dropdown-content` can catch and call this meth
 ```
 
-### d2l-dropdown-tabs
+### Content: Tabs [d2l-dropdown-tabs]
````
Also h2

annabelsegalld2l commented on components/dropdown/dropdown-content.js:
```diff
@@ -9,6 +9,7 @@ import { dropdownContentStyles } from './dropdown-content-styles.js';
  * @slot footer - Sticky container at the bottom of the dropdown
  * @fires d2l-dropdown-open - Dispatched when the dropdown is opened
  * @fires d2l-dropdown-close - Dispatched when the dropdown is closed
+ * @fires d2l-dropdown-focus-enter - Dispatched when user focus enters the dropdown content
```
Can we have some way to specify that this event only fires when the trap-focus option is on? Maybe just a "(trap-focus option only)" at the end of the line. Otherwise consumers might think that it happens whenever the user clicks inside the dropdown.

annabelsegalld2l commented on components/dropdown/README.md:
```diff
@@ -21,190 +29,230 @@
+  });
+</script>
 <d2l-dropdown-button text="Open!" primary>
   <d2l-dropdown-content min-width="150" max-width="400">
```
Do we want to add a mobile tray (ie. mobile-tray="bottom") to display the trays somewhere on this page? I was thinking to maybe add one of each tray as a demo, but putting it in with this example should be fine too. Not sure what the default size is for the snippets but it would activate at 616px.

GeorgePlukov replied:
I added it as default enabled within the live demo, it won't show until you resize the demos but once you hit `616px` it does appear

margaree commented on components/dropdown/README.md:
```diff
@@ -21,190 +29,230 @@
 
 | Property | Type | Description |
-|--|--|--|
+|---|---|
```
This table isn't actually working (I think it needs another `|---|`)

margaree commented on components/dropdown/README.md:
```diff
@@ -21,190 +29,230 @@
-
-**Properties:**
+<!-- docs: start hidden content -->
+### Properties:
```
```suggestion
### Properties
```

margaree commented on components/dropdown/README.md:
```diff
@@ -21,190 +29,230 @@
 
 If the dropdown is initially empty when it's opened, the dropdown pointer will not be positioned correctly.  In such cases, the `no-auto-open` attribute may be added to the opener, enabling you to take control of when the dropdown is actually opened:
```
This sentence becomes a bit strange without the code below

GeorgePlukov replied:
I changed the `:`  to a `.`, I think it makes sense to keep this line in here. Do you think it would be useful to add a code example back?

margaree replied:
I think so, though I think you could combine the two code chunks, e.g.:
```html
<script>

dropdown.addEventListener('click', function() {
  // fetch some content
  ...
  // take control of when the dropdown is actually opened
  dropdown.toggleOpen();
</script>
<d2l-dropdown no-auto-open>
  ...
</d2l-dropdown>
```

margaree commented on components/dropdown/README.md:
```diff
@@ -21,190 +29,230 @@
+  window.addEventListener('load', function () {
+    var dropdown = document.querySelector('d2l-dropdown-button');
+	setTimeout(() => {
+		dropdown.toggleOpen();
```
This looks to be using a mix of tabs and spaces

margaree commented on components/dropdown/README.md:
```diff
@@ -1,16 +1,24 @@
   import '@brightspace-ui/core/components/dropdown/dropdown.js';
   import '@brightspace-ui/core/components/dropdown/dropdown-content.js';
 </script>
-
```
Thoughts on hiding these code chunks?

GeorgePlukov commented:
> This is looking mostly good to go. Does it have the issue with jumpiness when the demos load after changing values in the properties table? I had that issue with date/time so I added a check for first load

Added that check to these too, it is a little bit annoying how it blinks open otherwise.

margaree commented on components/dropdown/README.md:
```diff
@@ -56,9 +81,11 @@ If the dropdown is initially empty when it's opened, the dropdown pointer will n
     setTimeout(() => {
-      dropdown.toggleOpen();
+      var dropdown = document.querySelector('d2l-dropdown-button');
+        dropdown.toggleOpen();
```
nit: spacing (as well as a few places below)

margaree commented on components/dropdown/README.md:
```diff
@@ -378,7 +415,10 @@ To make your usage of `d2l-dropdown-more` accessible, use the following property
+    const demoElem = document.querySelector('#demo-element');
+    if (!demoElem.hasAttribute('data-first-load')) return;
     setTimeout(() => {
+       var dropdown = document.querySelector('d2l-dropdown-button');
```
nit: dropdown is defined twice
